src/Core/Checkout/Refund/PaymentRefundProcessor.php

```diff
@@ -0,0 +1,101 @@
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\DataAbstractionLayer\EntityRepositoryInterface;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\Uuid\Uuid;
+
+class PaymentRefundProcessor
+{
+    /**
+     * @var OrderRefundStateHandler
+     */
+    private $refundStateHandler;
+
+    /**
+     * @var EntityRepositoryInterface
+     */
+    private $orderRefundRepository;
+
+    /**
+     * @var PaymentRefundHandlerRegistry
+     */
+    private $paymentRefundHandlerRegistry;
+
+    public function __construct(
+        EntityRepositoryInterface $orderRefundRepository,
+        PaymentRefundHandlerRegistry $paymentRefundHandlerRegistry,
+        OrderRefundStateHandler $refundStateHandler
+    ) {
+        $this->orderRefundRepository = $orderRefundRepository;
+        $this->paymentRefundHandlerRegistry = $paymentRefundHandlerRegistry;
+        $this->refundStateHandler = $refundStateHandler;
+    }
+
+    /**
+     * @throws InvalidOrderRefundException
+     * @throws PaymentRefundProcessException
+     */
+    public function processRefund(
+        string $orderRefundId,
+        Context $context
+    ): void {
+        if (!Uuid::isValid($orderRefundId)) {
+            throw new InvalidOrderRefundException($orderRefundId);
+        }
+
+        $criteria = new Criteria([$orderRefundId]);
+        $criteria->addAssociation('stateMachineState');
+        $criteria->addAssociation('paymentMethod');
+        $criteria->addAssociation('order');
+
+        /** @var OrderRefundEntity|null $orderRefund */
+        $orderRefund = $this->orderRefundRepository->search($criteria, $context)->first();
+
+        if (!$orderRefund) {
+            throw new InvalidOrderRefundException($orderRefundId);
+        }
+        $order = $orderRefund->getOrder();
+        if (!$order) {
+            throw new InvalidOrderException($orderRefund->getOrderId());
+        }
+        $stateMachineState = $orderRefund->getStateMachineState();
+        if (!$stateMachineState || $stateMachineState->getTechnicalName() !== OrderRefundStates::STATE_OPEN) {
+            throw new InvalidOrderRefundException($orderRefundId);
+        }
+
+        $paymentMethod = $orderRefund->getPaymentMethod();
+        if (!$paymentMethod) {
+            throw new UnknownPaymentMethodException($orderRefund->getPaymentMethodId());
+        }
+        $refundHandlerIdentifier = $paymentMethod->getRefundHandlerIdentifier();
+        if (!$refundHandlerIdentifier) {
+            throw new PaymentRefundUnsupportedException($orderRefund->getPaymentMethodId());
+        }
+
+        $paymentRefundHandler = $this->paymentRefundHandlerRegistry->getRefundHandler($refundHandlerIdentifier);
+        if (!$paymentRefundHandler) {
+            throw new PaymentRefundHandlerNotFoundException($refundHandlerIdentifier);
+        }
+
+        try {
+            $paymentRefundHandler->refund($orderRefundId, $context);
```

Comment:
why not putting the entity into the method? it is selected here already, and needs to be selected again in the Handler

Reply:
For public (service) methods it is generally preferred to only pass the id of an entity, because the entity might be modified by 3rd party code in the db or was not loaded with all the props the service might need. Thus the service can not guarantee the entities consistency with the db and that all necessary props are loaded and needs to (re-)fetch the entity anyway.

Reply:
yeah, I also thought about it again. some need more or less data we are fetching here. so this is okay :+1: